docs: reorganize docs to highlight tools and drivers (#4525)
apache/arrow-adbc: docs/source/driver/authoring.rst

Reviewer: Maybe we should be more explicit that you can build and host a driver wherever, and that the Foundry can provide hosting/distribution/etc?

Reviewer 2: Ok, I'll add that.

Reviewer 2: Done in 17e6c837466cb832313ed82b985bc6f6e66813bc

Writing New Drivers
===================

ADBC drivers can be written in C/C++, C#, Go, Java, or Rust.

> Because drivers expose a :doc:`standard C ABI <../format/specification>`, a driver written in any of these languages can be used from client libraries in any other language — so the implementation language is largely a matter of ecosystem fit and preference.

Reviewer: ```suggestion
Because drivers expose a :doc:`standard C ABI <../format/specification>`, a driver written in any of these languages can be used from client libraries in any other language, so the implementation language is largely a matter of ecosystem fit and preference.
```

.. list-table::
   :header-rows: 1

   * - Language
     - Documentation
     - Notes
   * - C/C++
     - :doc:`../cpp/index`
     - Drivers are loadable by the Driver Manager and usable from all client languages.
   * - C#
     - :doc:`../csharp/index`
     - Can experimentally export drivers for use from other languages via the C ABI.
   * - Go
     - `go/adbc <https://pkg.go.dev/github.com/apache/arrow-adbc/go/adbc>`__
     - Recommended for new drivers; easiest to package and distribute. Drivers are loadable by the Driver Manager and usable from all client languages.
   * - Java
     - :doc:`../java/index`
     - Drivers are usable from Java client libraries.
   * - Rust
     - :doc:`../rust/quickstart`
     - API definitions are still maturing. Drivers can be compiled to a C ABI shared library for use from other languages.

> Go is generally the recommended choice for new drivers.

Reviewer: ```suggestion
Go and Rust are generally the recommended choice for new drivers.
```

> C/C++ drivers have historically had dependency conflicts (e.g. with grpcio or pyarrow in Python processes), whereas Go libraries are easier to package and distribute without conflict.

Reviewer: ```suggestion
C/C++ drivers have historically had dependency conflicts (e.g. with grpcio or pyarrow in Python processes), whereas Go and Rust libraries are easier to package and distribute without conflict.
```

Writing New Drivers
===================

Interested in building an ADBC driver? The best place to start is the **ADBC
Driver Foundry**, a community project that hosts most ADBC drivers and provides
the frameworks, templates, testing tools, and distribution infrastructure for
building and shipping them.

.. button-link:: https://docs.adbc-drivers.org/building-drivers/
   :color: primary

   Building Drivers guide (ADBC Driver Foundry) :octicon:`link-external`

The rest of this page gives a brief orientation. For anything beyond
that—project templates, frameworks, validation tests, packaging, and
distribution—follow the guide above.

.. note::

   The ADBC project itself focuses on the :doc:`ADBC standard
   <../format/specification>` and :doc:`client libraries <../client_libraries>`,
   not on building and distributing drivers. In most cases, new drivers should
   be developed in the ADBC Driver Foundry, not in the ``apache/arrow-adbc``
   repository.

>    The ADBC Driver Foundry is an independent community project, separate from

Author: Are we avoiding the term third-party on purpose?

Reviewer 2: Not necessarily, but nor do I think it's the most informative word to describe the relationship between the Foundry and the ADBC project.
